dynamic_validators/insecure_dynamic_validators.md

```diff
@@ -0,0 +1,142 @@
+# Dynamically add/remove validators at runtime (Insecure version)
+
+## Problem Description
+In the current state of tendermint bdb, validator nodes are specified in the Tendermint config file before starting the node. It is necessary that the users should be able to dynamically add new validators.
+
+**NOTE**: This document presents an insecure implementation intended for bigchaindb tendermint MVP.
+
+### Technical details
+Tendermint allows the abci client application to return a list of validators during `end_block` call. If the client doesn't wish to modify the validators an empty list should be returned. The return value allows the client to dynamically add/remove validators. Note that the return value could be thought of as a (validators) diff of desired validators list and current validators list. Below is an example of validators diff,
```

Comment:
Is BigchainDB supposed to return a diff / delta / change-set or a _complete_ list of validators (in the End Block response)?

After reading and re-reading [the Tendermint docs about End Block](https://tendermint.readthedocs.io/en/master/app-development.html#endblock), I now think the answer is:

- If you don't want to change the validators (or their powers) at all, then return an empty list `[]` in the End Block response.
- If you _do_ want to change the validators in some way, then _return a complete list of all validators_ with any new ones added, and with the ones you want to delete having their power set to 0.

Reply:
@ttmc I don't think one needs to _return a complete list of all validators_ . The documentation says,
> To add a new validator or update an existing one, simply include them in the list returned

The way I interpret this is that in case a user doesn't wish to update an existing validator they shouldn't need to include them in the list. That is whey I refered to it has a diff. 

Am I missing something?

Reply:
> To add a new validator or update an existing one, simply include them in the list returned

That sentence doesn't say anything about whether you should or shouldn't include the unchanging set of validators as well. Honestly, I'm _not sure_. To find out the answer, I guess you could look at the code, ask in a Tendermint help channel, or just do lots of tests.

Reply:
@ttmc I looked in the tendermint code base [here](https://github.com/tendermint/tendermint/blob/master/state/execution.go#L251). I seems that the validator set `currentSet` is modified depending on the updates received. It seems that we don't need to return all the validators, rathers just the ones which we wish to add, modify or delete.